packages/core/src/StateNode.ts

```diff
@@ -956,8 +956,8 @@ class StateNode<
     const reentryNodes: StateNode<any, any, any, any, any>[] = [];
 
     if (!isInternal) {
-      allNextStateNodes.forEach((n1) => {
-        reentryNodes.push(...this.getExternalReentryNodes(n1));
+      nextStateNodes.forEach((targetNode) => {
```

Comment:
Note the change here from `allNextStateNodes` to `nextStateNodes`.

`nextStateNodes` represent the target state nodes literally used in the transition definition whereas `allNextStateNodes` represent "resolved" targets (initial states etc already resolved). To compute correctly what has to be reentered we need to consider the target states before resolving them.